### PR TITLE
feat(clai2): store Codex tokens in a 0600 file when no keyring backend exists

### PR DESCRIPTION
**Pull request by mpfaffenberger**
## Why

On a headless Linux box or over SSH, `keyring` typically resolves to its `fail` backend, which raises `NoKeyringError` on every call. That made `/login openai-codex` (and every subsequent token refresh) blow up with no way forward.

## What

- Keyring is still the first choice and unchanged (chunking, Windows size limits, readback verification all as in #886).
- Only `NoKeyringError` and `InitError` count as "no keyring". On those, tokens go to `credentials.json` next to `config.db`, written via a `0600` temp file plus `os.replace` so a mid-refresh crash cannot leave half a bundle. The staging name is unique per write (`credentials.json.<uuid4>.tmp`) and opened `O_CREAT|O_EXCL`, so a symlink or file already at that path is refused rather than followed, concurrent saves cannot collide, and `0600` always applies to a file we created. Staging leftovers from a killed write are swept before each save. A locked keyring (`KeyringLocked`) still raises; we do not silently downgrade someone who has a keyring.
- Load order: keyring value, else the file. A successful keyring save deletes the file, so the moment a keyring becomes available the plaintext copy disappears on the next login or refresh.
- The fallback file is derived per account (`credentials-<account>.json`): `credentials-openai-codex.json`, `credentials-vllm.json`, `credentials-openrouter.json`. main's credential store is now shared by the vllm and openrouter connections, so they get the fallback too, with no changes at their call sites.
- `/login` says where tokens went: the "saved in plaintext at <path>" wording is only produced when the file exists after the save.
- `credentials_path()` owns the one XDG rule (mirroring `SettingsStore`'s directory, which cannot be imported without constructing a database). Callers can pass an explicit `fallback=` to override it, which is what the tests do.

## Verified

- ruff / pyright strict / 158 tests, 100% branch coverage. New tests cover both trigger exceptions, the locked-keyring non-fallback, file-to-keyring migration, mode `0600`, no leftover `.tmp`, and the login message.
- Ran the real `keyring.backends.fail.Keyring` (no monkeypatching): save/load round-trips through the file with `0o600`, with symlinks planted at the staging paths leaving the target untouched and no token-bearing leftovers behind. Repeated for the vllm and openrouter accounts to confirm each gets its own `0600` file.

README's Codex authentication section and the PLUGINS.md credentials note are updated; the latter previously promised there was no plaintext fallback.

### CODE REVIEW

File: pydantic-clai2/src/pydantic_clai2/credential_store.py [hunk 1]
```diff
@@ -41,7 +45,38 @@ def _delete(*, services: list[str]) -> None:
             pass  # A failed write may not have created the entry.
 
 
-def load_codex_credentials() -> str | None:
+def _write_private(*, path: Path, value: str) -> None:
+    """Replace the file atomically so a refresh cannot leave half a token bundle behind."""
+    path.parent.mkdir(parents=True, exist_ok=True)
+    staging = path.with_name(f'{path.name}.tmp')
+    descriptor = os.open(staging, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
```
[the remark this answers is not in the document]
Comment (mpfaffenberger) [reply pydantic-clai2/src/pydantic_clai2/credential_store.py]: Fixed in 92d311f6, exactly as suggested: unique `credentials.json.<uuid4>.tmp` staging name opened `O_CREAT|O_EXCL`. Nothing already at the path is ever written through, and because the file is always freshly created, `0600` now applies to a file we made rather than one an attacker pre-created with looser mode. `finally` unlinks the staging path so a failed or cancelled write leaves nothing behind.

File: pydantic-clai2/src/pydantic_clai2/credential_store.py [hunk 2]
```diff
@@ -41,7 +45,38 @@ def _delete(*, services: list[str]) -> None:
             pass  # A failed write may not have created the entry.
 
 
-def load_codex_credentials() -> str | None:
+def _write_private(*, path: Path, value: str) -> None:
+    """Replace the file atomically so a refresh cannot leave half a token bundle behind."""
+    path.parent.mkdir(parents=True, exist_ok=True)
+    staging = path.with_name(f'{path.name}.tmp')
```
[the remark this answers is not in the document]
Comment (mpfaffenberger) [reply pydantic-clai2/src/pydantic_clai2/credential_store.py]: Fixed in 92d311f6: the staging name is unique per write (`credentials.json.<uuid4>.tmp`), so concurrent saves no longer share one entry and one process's `os.replace` cannot remove another's staging file. Same root cause as the symlink issue, so one change covers both.

File: pydantic-clai2/src/pydantic_clai2/credential_store.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpfaffenberger) [reply pydantic-clai2/src/pydantic_clai2/credential_store.py]: Fixed in 92d311f6: the fixed `.tmp` path is gone. Staging is now `credentials.json.<uuid4>.tmp` opened with `O_CREAT|O_EXCL`, which refuses anything already there instead of following a symlink or inheriting a permissive mode, and the always-newly-created file guarantees `0600`. Stale staging files from a killed write are swept before saving so they cannot accumulate tokens, and the staging path is unlinked in a `finally` on failure.

**Comment by mpfaffenberger** [comment]
Both review bots flagged the same real bug in the fallback writer, and it was mine: staging reused a predictable `credentials.json.tmp` opened with `O_CREAT|O_TRUNC`, so

- a symlink planted at that path received the OAuth tokens,
- a pre-existing file kept its own (possibly permissive) mode instead of `0600`,
- concurrent saves shared one staging entry, so one process's `os.replace` broke another's.

Fixed in `92d311f6` with a unique `credentials.json.<uuid4>.tmp` name opened `O_CREAT|O_EXCL` (never writes through anything already at the path), a `finally` unlink, and a sweep of stale staging files from a killed write so tokens cannot accumulate in them.

New tests cover the planted-symlink path, a simulated TOCTOU race where the file appears between the sweep and the open, and stale-staging cleanup. Reproduced the attack against the real `keyring.backends.fail.Keyring` with symlinks planted at the staging paths: target file untouched, credentials written `0600`, no token-bearing leftovers.

`test on 3.10 (all-extras)` was red only on the unrelated flaky `tests/shell/test_persistent.py::test_supervisor_killed_mid_command_returns_stale_status` ("Shell supervisor exited with -15"); it passes on main, so I reran the failed jobs rather than touching it.

File: pydantic-clai2/tests/test_credential_store.py [hunk 3]
```diff
@@ -135,5 +156,123 @@ def fail_after_write(service: str, account: str, value: str) -> None:
...
+    fallback.write_text('{"access_token":"from-file"}')
+    assert load_codex_credentials(fallback=fallback) == '{"access_token":"from-file"}'
+    save_codex_credentials(fallback=fallback, value='{"access_token":"in-keyring"}')
+    assert not fallback.exists()
+    assert load_codex_credentials(fallback=fallback) == '{"access_token":"in-keyring"}'
+
+
+async def test_login_reports_plaintext_location(no_keyring: None, monkeypatch: pytest.MonkeyPatch) -> None:
+    """The fallback path is derived from the account, so login can name the file it wrote."""
+    credentials = OpenAICodexCredentials(access_token='a', refresh_token='r', account_id='x')
```
[the remark this answers is not in the document]
Comment (mpfaffenberger) [reply pydantic-clai2/tests/test_credential_store.py:253]: This one does not reproduce: `tests/conftest.py` has an autouse `isolated_environment` fixture that sets `XDG_CONFIG_HOME` to `tmp_path/config` and `HOME` to `tmp_path/home` for every test, and `credentials_path()` reads `XDG_CONFIG_HOME` at call time. Verified by running the test alone and checking that `~/.config/pydantic-clai2/credentials-openai-codex.json` does not exist before or after. Adding a second `setenv` in the test would only duplicate the fixture.

File: pydantic-clai2/src/pydantic_clai2/credential_store.py [hunk 4]
```diff
@@ -14,6 +16,18 @@
...
+_NO_KEYRING = (NoKeyringError, InitError)
+
+
+def credentials_path(*, account: str = _ACCOUNT) -> Path:
+    """The private fallback file for one account, beside `config.db`.
+
+    Mirrors `SettingsStore`'s directory rule, which cannot be imported here without
+    constructing a database, and keeps accounts apart so one login cannot overwrite another.
+    """
+    root = Path(os.getenv('XDG_CONFIG_HOME', str(Path.home() / '.config'))) / 'pydantic-clai2'
```
[the remark this answers is not in the document]
Comment (mpfaffenberger) [reply pydantic-clai2/src/pydantic_clai2/credential_store.py:30]: Fair on the docs, wrong on the behaviour. `--database PATH` selects a settings database; it has never selected where credentials live, because keyring entries are per user and ignore it too. The file fallback keeps that: per user, in `$XDG_CONFIG_HOME/pydantic-clai2/`, whichever database is open. Threading the store path through would also have to reach `vllm.py` and `openrouter.py` module functions and would make credentials silently diverge between profiles in a way keyring never did. Fixed the README, PLUGINS.md, and the `credentials_path` docstring to say "per user, `--database` does not move them" instead of "next to `config.db`".
